oci/caps/utils.go

```diff
@@ -8,63 +8,35 @@ import (
 	"github.com/syndtr/gocapability/capability"
 )
 
-var capabilityList Capabilities
+var (
+	allCaps []string
 
-func init() {
-	last := capability.CAP_LAST_CAP
-	// hack for RHEL6 which has no /proc/sys/kernel/cap_last_cap
-	if last == capability.Cap(63) {
-		last = capability.CAP_BLOCK_SUSPEND
-	}
-	for _, cap := range capability.List() {
-		if cap > last {
-			continue
-		}
-		capabilityList = append(capabilityList,
-			&CapabilityMapping{
-				Key:   "CAP_" + strings.ToUpper(cap.String()),
-				Value: cap,
-			},
-		)
-	}
-}
-
-type (
-	// CapabilityMapping maps linux capability name to its value of capability.Cap type
+	// capabilityList maps linux capability name to its value of capability.Cap
+	// type. This list contains nil entries for capabilities that are known, but
+	// not supported by the current kernel.
 	// Capabilities is one of the security systems in Linux Security Module (LSM)
 	// framework provided by the kernel.
 	// For more details on capabilities, see http://man7.org/linux/man-pages/man7/capabilities.7.html
-	CapabilityMapping struct {
-		Key   string         `json:"key,omitempty"`
-		Value capability.Cap `json:"value,omitempty"`
-	}
-	// Capabilities contains all CapabilityMapping
-	Capabilities []*CapabilityMapping
+	capabilityList map[string]*capability.Cap
 )
 
-// String returns <key> of CapabilityMapping
-func (c *CapabilityMapping) String() string {
-	return c.Key
-}
-
-// GetCapability returns CapabilityMapping which contains specific key
-func GetCapability(key string) *CapabilityMapping {
-	for _, capp := range capabilityList {
-		if capp.Key == key {
-			cpy := *capp
-			return &cpy
+func init() {
+	allCaps = make([]string, capability.CAP_LAST_CAP+1)
+	capabilityList = make(map[string]*capability.Cap, len(capability.List()))
+	for i, c := range capability.List() {
```

Comment:
The implementation of `capability.List()` appears to generate/return a new slice of the full list, so it seems a tiny bit wasteful to call twice -- should probably go into a variable for `len()` vs `for`?

```suggestion
	allCaps = make([]string, capability.CAP_LAST_CAP+1)
	rawCaps := capability.List()
	capabilityList = make(map[string]*capability.Cap, len(rawCaps))
	for i, c := range rawCaps {
```

Does it also make sense to use `len(rawCaps)` for the size of `allCaps`, given that we do `allCaps[i] = ...` below, which will be based on the number of capabilities total, not on the maximum value?  (or is that an implementation bug that should've been `allCaps[c]` instead?  Looking at how it gets used/returned, it doesn't seem like a bug so it looks like we're allocating more space than we actually need here. :sweat_smile:)

```suggestion
	rawCaps := capability.List()
	allCaps = make([]string, len(rawCaps))
	capabilityList = make(map[string]*capability.Cap, len(rawCaps))
	for i, c := range rawCaps {
```

Reply:
> Looking at how it gets used/returned, it doesn't seem like a bug so it looks like we're allocating more space than we actually need here

Hmmm good point, so I don't think it's a bug (but I do take advantage of the fact that all the caps are in sequential order as proxy for "length") as `CAP_LAST_CAP` should be `<= len(capability.List()`;

https://github.com/moby/moby/blob/9c15e82f19b0ad3c5fe8617a8ec2dddc6639f40a/vendor/github.com/syndtr/gocapability/capability/capability_linux.go#L52

Hmmm... although

- it's initially set to `Cap(63)`, and will be if the code above check fails
- newer kernels may support more capabilities than what the library supports, so we should check something like `math.Min(len(rawCaps), CAP_LAST_CAP+1)`